src/IO/ForkWriteBuffer.cpp

```diff
@@ -0,0 +1,83 @@
+#include <IO/ForkWriteBuffer.h>
+#include <Common/Exception.h>
+#include <ranges>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_CREATE_IO_BUFFER;
+}
+
+ForkWriteBuffer::ForkWriteBuffer(WriteBufferPtrs && sources_)
+    : WriteBuffer(nullptr, 0), sources(std::move(sources_))
+{
+    if (sources.empty())
+    {
+        first_buffer = nullptr;
+        throw Exception("ForkWriteBuffer required WriteBuffer is not provided", ErrorCodes::CANNOT_CREATE_IO_BUFFER);
+    }
+    else
+    {
+        first_buffer = sources.begin()->get();
+        set(first_buffer->buffer().begin(), first_buffer->buffer().size());
+    }
+}
+
+
+void ForkWriteBuffer::nextImpl()
+{
+    if (!first_buffer)
+        return;
```

Comment:
If we make sure that this buffer cannot be created with an empty set of buffers (there is exception in constructor now), then this check is not needed.

Reply:
Also probably there is no point to have a separate variable `first_buffer` as it always equals `sources.front()`. In my opinion it adds redundant complexity with worrying that `first_buffer`, which is a raw pointer, is valid.

Reply:
good catch! i have removed `first_buffer`